src/machine_charm.py

```diff
@@ -149,11 +149,14 @@ class GrafanaAgentMachineCharm(GrafanaAgentCharm):
     service_name = "grafana-agent.grafana-agent"
 
     mandatory_relation_pairs = [
-        ("cos-agent", ["send-remote-write", "grafana-cloud-config"]),
-        ("cos-agent", ["logging-consumer", "grafana-cloud-config"]),
-        ("cos-agent", ["grafana-dashboards-provider", "grafana-cloud-config"]),
-        ("juju-info", ["send-remote-write", "grafana-cloud-config"]),
-        ("juju-info", ["logging-consumer", "grafana-cloud-config"]),
+        (
+            "cos-agent",
+            (
+                ["grafana-cloud-config"],
+                ["send-remote-write", "logging-consumer", "grafana-dashboards-provider"],
+            ),
+        ),
+        ("juju-info", (["grafana-cloud-config"], ["send-remote-write", "logging-consumer"])),
```

Comment:
I think some comments could make this very clear:
```suggestion
        (
            "cos-agent",  # must be paired with:
            (
                ["grafana-cloud-config"],  # or
                ["send-remote-write", "logging-consumer", "grafana-dashboards-provider"],
            ),
        ),
        (
            "juju-info",  # must be paired with:
            (
                ["grafana-cloud-config"],  # or
                ["send-remote-write", "logging-consumer"]
            )
        ),
```

Reply:
- Do we need the external pair of `()`?
- Should it be a dict? `Dict[str, List[Set[str]]]`

```python
mandatory_relation_pairs = {
    "cos-agent": [  # must be paired with:
        {"grafana-cloud-config"},  # or
        {"send-remote-write", "logging-consumer", "grafana-dashboards-provider"},
    ],
}
```

Reply:
But in reality it would rather be

```
mandatory_relation_pairs = {
    "cos-agent": [  # must be paired with:
        {"grafana-cloud-config"},  # or
        {"send-remote-write", "grafana-dashboards-provider"},
        {"logging-consumer", "grafana-dashboards-provider"},
    ],
}
```

right?

Reply:
Ah, nvmd. Looking closer I got it.